internal/audit/kafka/publisher.go

```diff
@@ -0,0 +1,187 @@
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Publisher{
+		Client:         client,
+		async:          conf.Async,
+		decisionFilter: decisionFilter,
+		marshaller:     recordMarshaller{Encoding: conf.Encoding},
+	}, nil
+}
+
+func (p *Publisher) Close() {
+	p.Client.Close()
+}
+
+func (p *Publisher) Backend() string {
+	return Backend
+}
+
+func (p *Publisher) Enabled() bool {
+	return true
+}
+
+func (p *Publisher) WriteAccessLogEntry(ctx context.Context, record audit.AccessLogEntryMaker) error {
+	rec, err := record()
+	if err != nil {
+		return err
+	}
+
+	msg, err := p.marshaller.MarshalAccessLogEntry(rec)
+	if err != nil {
+		return err
+	}
+
+	return p.write(ctx, msg)
+}
+
+func (p *Publisher) WriteDecisionLogEntry(ctx context.Context, record audit.DecisionLogEntryMaker) error {
+	rec, err := record()
+	if err != nil {
+		return err
+	}
+
+	if p.decisionFilter != nil {
+		rec = p.decisionFilter(rec)
+		if rec == nil {
+			return nil
+		}
+	}
+
+	msg, err := p.marshaller.MarshalDecisionLogEntry(rec)
+	if err != nil {
+		return err
+	}
+
+	return p.write(ctx, msg)
+}
+
+func (p *Publisher) write(ctx context.Context, msg *kgo.Record) error {
+	if !p.async {
+		return p.Client.ProduceSync(ctx, msg).FirstErr()
+	}
+
+	p.Client.Produce(ctx, msg, func(r *kgo.Record, err error) {
+		if err != nil {
+			// TODO: Handle via interceptor
+			ctxzap.Extract(ctx).Warn("failed to write audit log entry", zap.Error(err))
+		}
+	})
+	return nil
+}
+
+type recordMarshaller struct {
+	Encoding string
+}
+
+func (m recordMarshaller) MarshalAccessLogEntry(rec *auditv1.AccessLogEntry) (*kgo.Record, error) {
+	callID, err := audit.ID(rec.CallId).Repr()
+	if err != nil {
+		return nil, fmt.Errorf("invalid call ID: %w", err)
+	}
+
+	var payload []byte
+	switch m.Encoding {
+	default:
+		return nil, fmt.Errorf("invalid encoding format: %s", m.Encoding)
+	case EncodingJSON:
+		payload, err = protojson.Marshal(rec)
+	case EncodingProtobuf:
+		payload, err = rec.MarshalVT()
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal entry: %w", err)
+	}
+
+	return m.record(callID.Bytes(), payload)
+}
+
+func (m recordMarshaller) MarshalDecisionLogEntry(rec *auditv1.DecisionLogEntry) (*kgo.Record, error) {
+	callID, err := audit.ID(rec.CallId).Repr()
+	if err != nil {
+		return nil, fmt.Errorf("invalid call ID: %w", err)
+	}
+
+	var payload []byte
+	switch m.Encoding {
+	default:
+		return nil, fmt.Errorf("invalid encoding format: %s", m.Encoding)
+	case EncodingJSON:
+		payload, err = protojson.Marshal(rec)
+	case EncodingProtobuf:
+		payload, err = rec.MarshalVT()
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal entry: %w", err)
+	}
+
+	return m.record(callID.Bytes(), payload)
+}
+
+func (m recordMarshaller) record(key, payload []byte) (*kgo.Record, error) {
+	return &kgo.Record{
+		Key:   key,
+		Value: payload,
+		Headers: []kgo.RecordHeader{
+			{
+				Key:   encodingHeaderKey,
+				Value: []byte(m.Encoding),
+			},
+		},
```

Comment:
It might be helpful to include the kind of record (access or decision) as a header as well.